cpp/src/utilities/type_dispatcher.hpp

```diff
@@ -107,14 +112,83 @@ decltype(auto) type_dispatcher(gdf_dtype dtype,
     case GDF_DATE64:    { return f.template operator()< date64 >(std::forward<Ts>(args)...); }
     case GDF_TIMESTAMP: { return f.template operator()< timestamp >(std::forward<Ts>(args)...); }
     case GDF_CATEGORY:  { return f.template operator()< category >(std::forward<Ts>(args)...); }
-    default:            { assert(false && "type_dispatcher: invalid gdf_type"); }
+    default: { assert(false && "type_dispatcher: invalid gdf_type"); }
   }
   // Need to find out what the return type is in order to have a default return value
   // and solve the compiler warning for lack of a default return
   using return_type = decltype(f.template operator()<int8_t>(std::forward<Ts>(args)...));
   return return_type();
 }
 
-} // namespace cudf
+/**---------------------------------------------------------------------------*
+ * @brief Type-traits struct for mapping a C++ type to it's corresponding
+ * gdf_dtype.
+ *
+ * This traits structure contains a `gdf_dtype` member `value` that returns the
+ * appropriate `gdf_dtype` for the specified C++ type.
+ *
+ * @tparam T The type to map to a `gdf_dtype`
+ *---------------------------------------------------------------------------**/
+template <typename T>
+struct type_to_gdf_dtype {
+  static constexpr gdf_dtype value{GDF_invalid};
+};
+
+template <>
+struct type_to_gdf_dtype<int8_t> {
+  static constexpr gdf_dtype value{GDF_INT8};
+};
+
+template <>
+struct type_to_gdf_dtype<int16_t> {
+  static constexpr gdf_dtype value{GDF_INT16};
+};
+
+template <>
+struct type_to_gdf_dtype<int32_t> {
+  static constexpr gdf_dtype value{GDF_INT32};
+};
+
+template <>
+struct type_to_gdf_dtype<int64_t> {
+  static constexpr gdf_dtype value{GDF_INT64};
+};
+
+template <>
+struct type_to_gdf_dtype<float> {
+  static constexpr gdf_dtype value{GDF_FLOAT32};
+};
+
+template <>
+struct type_to_gdf_dtype<double> {
+  static constexpr gdf_dtype value{GDF_FLOAT64};
+};
+
+template <>
+struct type_to_gdf_dtype<cudf::date32> {
+  static constexpr gdf_dtype value{GDF_DATE32};
+};
+
+template <>
+struct type_to_gdf_dtype<cudf::date64> {
+  static constexpr gdf_dtype value{GDF_DATE64};
+};
+
+template <>
+struct type_to_gdf_dtype<cudf::timestamp> {
+  static constexpr gdf_dtype value{GDF_TIMESTAMP};
+};
+
+template <>
+struct type_to_gdf_dtype<cudf::category> {
+  static constexpr gdf_dtype value{GDF_CATEGORY};
+};
+
+template <>
+struct type_to_gdf_dtype<NVStrings> {
+  static constexpr gdf_dtype value{GDF_STRING};
+};
+
+}  // namespace cudf
```

Comment:
I don't think you need such a long specialization... this is a constexpr function and we're in C++14. You should be able to use an appropriate tuple of the relevant types, and convert indices to tuple component element types or back and forth.

Of course, that's just prettying up the code and removing a bit of redundancy; if this works, it works.

By the way - I'm a fan of enums instead of static constexpr's, because they're guaranteed never to actually have storage in memory and only be immediate values, but many people disagree. Still, you need to make it `static constexpr const` I think.

Reply:
It would appear there's no performance/optimization difference. The only difference is that `static constexpr` members can't be ODR-used.

https://isocpp.org/blog/2016/05/quick-q-static-constexpr-int-vs-old-fashioned-enum-when-and-why

However, this is no longer true in C++17:

https://en.cppreference.com/w/cpp/language/static

> If a const non-inline (since C++17) static data member or a constexpr static data member (since C++11) is odr-used, a definition at namespace scope is still required, but it cannot have an initializer. **This definition is deprecated for constexpr data members (since C++17).**

I find the `static constexpr` members to be more intuitive, especially for those who've never seen the member `enum` pattern before.

Reply:
The FAQ is only right up until C++17, at which point the program it lists will [compile just fine](https://godbolt.org/) and _allocate memory space_ for the `static constexpr`.

And the worst thing is, you as the author of the class/struct can't _prevent_ the using code from taking the address-of, or a reference-to, that variable.

So despite the aesthetic point, about which I agree - for now I'm sticking to enums to make sure my code doesn't slip from under my figners.

Reply:
I agree that this could be done w/o the explicit list of specializations, but in the interest of time I'm going to leave it as-is. If I have an abundance of free time and am feeling motivated later on, I'll update it.